Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -839,18 +839,28 @@ extension TaskHandler: ChannelDuplexHandler {
         }.flatMap {
             self.writeBody(request: request, context: context)
         }.flatMap {
-            self.state = .bodySent
             context.eventLoop.assertInEventLoop()
-            if let expectedBodyLength = self.expectedBodyLength, expectedBodyLength != self.actualBodyLength {
-                let error = HTTPClientError.bodyLengthMismatch
-                self.errorCaught(context: context, error: error)
-                return context.eventLoop.makeFailedFuture(error)
+            switch self.state {
+            case .endOrError:
+                return context.eventLoop.makeSucceededFuture(())
+            default:
+                self.state = .bodySent
+                if let expectedBodyLength = self.expectedBodyLength, expectedBodyLength != self.actualBodyLength {
+                    let error = HTTPClientError.bodyLengthMismatch
+                    self.errorCaught(context: context, error: error)
```

Comment:
We don't need `errorCaught` here: the future chain will call this code path anyway.

Reply:
fixed, thanks!